src/main/java/org/elasticsearch/index/query/TermsFilterParser.java

```diff
@@ -188,7 +188,12 @@ public Filter parse(QueryParseContext parseContext) throws IOException, QueryPar
 
             // external lookup, use it
             TermsLookup termsLookup = new TermsLookup(lookupIndex, lookupType, lookupId, lookupRouting, lookupPath, parseContext, SearchContext.current());
-            terms.addAll(termsFilterCache.terms(termsLookup, lookupCache, cacheKey));
+            // We always pass null as the cacheKey here:  Using the cache key for the overall filter generally leads
+            // to redundant lookup caching:  The overall filter needs to be keyed off the field being filtered +
+            // the lookup options.  Whereas the lookup only needs to be keyed off the lookup options.
+            // Note that we avoid a copy here and just assign to the looked-up terms.  This is safe so long as
+            // terms is never modified.
+            terms = termsFilterCache.terms(termsLookup, lookupCache, null);
```

Comment:
As described in the comment, I don't pass through the cache key (the last null argument) so as not to duplicate the terms lookups.